Security/EditableRolesBuilder.php

```diff
@@ -85,43 +105,98 @@ public function getRoles()
+            if (in_array($id, $this->exclude)) {
                 continue;
             }
 
-            foreach ($admin->getSecurityInformation() as $role => $permissions) {
-                $role = sprintf($baseRole, $role);
-
-                if ($isMaster) {
-                    // if the user has the MASTER permission, allow to grant access the admin roles to other users
-                    $roles[$role] = $role;
-                } elseif ($this->authorizationChecker->isGranted($role)) {
-                    // although the user has no MASTER permission, allow the currently logged in user to view the role
-                    $rolesReadOnly[$role] = $role;
+            $isMaster = ($admin->isGranted('MASTER') || $admin->isGranted('OPERATOR') ||
+                $this->authorizationChecker->isGranted('ROLE_SUPER_ADMIN') ||
+                $this->authorizationChecker->isGranted('ROLE_SONATA_ADMIN'));
+            $securityHandler = $admin->getSecurityHandler();
+            $baseRole = $securityHandler->getBaseRole($admin);
+            $groupPermission = $admin->getSecurityInformation();
+            $this->labelPermission = array_keys($groupPermission);
+            $this->labelAdmin[] = $admin->trans($admin->getLabel());
+
+            foreach ($groupPermission as $role => $permissions) {
+                $roles[str_replace('.', '_', $id)][sprintf($baseRole, $role)] = $role;
+                if (!$isMaster) {
+                    $rolesReadOnly[] = sprintf($baseRole, $role);
                 }
             }
         }
 
-        $isMaster = $this->authorizationChecker->isGranted('ROLE_SUPER_ADMIN');
+        $roles['other'] = array(
+            'ROLE_ADMIN' => 'Role Admin',
+            'ROLE_SUPER_ADMIN' => 'Role Super Admin',
+            'ROLE_SONATA_ADMIN' => 'Role Sonata Admin',
+        );
 
-        // get roles from the service container
         foreach ($this->rolesHierarchy as $name => $rolesHierarchy) {
             if ($this->authorizationChecker->isGranted($name) || $isMaster) {
-                $roles[$name] = $name.': '.implode(', ', $rolesHierarchy);
-
                 foreach ($rolesHierarchy as $role) {
-                    if (!isset($roles[$role])) {
-                        $roles[$role] = $role;
+                    if (false === array_key_exists($role, $this->rolesHierarchy)
+                        && !isset($roles['other'][$role]) && false === $this->recursiveArraySearch($role, $roles)) {
+                        $roles['other'][$role] = ucfirst(mb_strtolower(str_replace('_', ' ', $role)));
                     }
                 }
             }
         }
 
+        if (empty($this->labelPermission)) {
+            throw new \InvalidArgumentException('You must add this line in the configuration of Sonata Admin: '.
+                "[security:\n\thandler: sonata.admin.security.handler.role]");
+        }
```

Comment:
This could be, if requires to do additional config to use the bundle like before